### PR TITLE
Chambering Refactor and Single Action Revolvers

### PR DESCRIPTION
**Pull request by hyper2snyper**
## About The Pull Request

Minor Cock() refactor. Mainly just cleaning up the unique_action() bloat.
Adds code for single-action revolvers. Have fun MJP

## Why It's Good For The Game

Pew Pew Clint Eastwood style

## Changelog
:cl:
code: Single Action Revolver Support
refactor: Cock() Refactor and removes some Unique_Action() bloat
/:cl:

### CODE REVIEW

File: code/modules/projectiles/guns/revolvers.dm [hunk 1]
```diff
@@ -536,3 +534,48 @@
...
+/obj/item/weapon/gun/revolver/single_action/able_to_fire(mob/user)
+	. = ..()
+	if(. && !primed) //If the chamber is open theres no need to give the unprimed message.
+		to_chat(user, span_warning("You need to prime the revolver first!"))
+		return FALSE
+
+/obj/item/weapon/gun/revolver/single_action/Fire()
+	. = ..()
+	primed = FALSE
+	update_icon_state()
```
Comment (SplinterGP) [inline code/modules/projectiles/guns/revolvers.dm:564]: what ?

**Review by SplinterGP — CHANGES_REQUESTED** [review]

File: code/modules/projectiles/guns/revolvers.dm [hunk 2]
```diff
@@ -536,3 +534,48 @@
 	burst_delay = 0.1 SECONDS
 	scatter_unwielded = 20
 	damage_mult = 1.05
+
+/obj/item/weapon/gun/revolver/single_action //This town aint big enuf fer the two of us
+	name = "single action revolver"
+	desc = "you should not be seeing this."
+	current_mag = /obj/item/ammo_magazine/internal/revolver/m44
+	///Whether or not the revolvers hammer is ready to fire.
+	var/primed = FALSE
```
Comment (SplinterGP) [inline code/modules/projectiles/guns/revolvers.dm:543]: snowflake ?

**Review by SplinterGP — CHANGES_REQUESTED** [review]

**Review by SplinterGP — CHANGES_REQUESTED** [review]
don't create new snowflake stuff, use the already existing guns proc that reloads, chambers, shoots and spins the barrel to next position to do it, overriding if necessary/adding a flagcheck on fire() if it should chamber next round automatically or not.

File: code/modules/projectiles/gun_system.dm [hunk 3]
```diff
@@ -364,7 +364,8 @@
 /obj/item/weapon/gun/unique_action(mob/user)
 	. = ..()
 	if(CHECK_BITFIELD(flags_item, IS_DEPLOYABLE) && !CHECK_BITFIELD(flags_item, IS_DEPLOYED)) //If the gun can be deployed, it deploys when unique_action is called.
-		return FALSE
+		return
```
Comment (TiviPlus) [inline code/modules/projectiles/gun_system.dm:367]: return whatever cock returns when failing here I guess?

File: code/modules/projectiles/guns/revolvers.dm (hunk 1 above)
Comment (TiviPlus) [reply code/modules/projectiles/guns/revolvers.dm:564]: use update icon

**Review by TiviPlus — CHANGES_REQUESTED** [review]

File: code/modules/projectiles/guns/revolvers.dm [hunk 4]
```diff
@@ -55,16 +62,6 @@
...
 	current_mag.chamber_position = current_mag.chamber_position == 1 ? current_mag.max_rounds : current_mag.chamber_position - 1
 
-/obj/item/weapon/gun/revolver/proc/spin_cylinder(mob/user)
-	if(!current_mag.chamber_closed) //We're not spinning while it's open. Could screw up reloading.
-		return FALSE
-	current_mag.chamber_position = rand(1,current_mag.max_rounds)
-	to_chat(user, span_notice("You spin the cylinder."))
-	playsound(user, cocked_sound, 25, 1)
-	russian_roulette = !russian_roulette //Sets to play RR. Resets when the gun is emptied.
-	return TRUE
```
Comment (SplinterGP) [inline code/modules/projectiles/guns/revolvers.dm:65]: why did you remove this.

**Review by SplinterGP — CHANGES_REQUESTED** [review]

File: code/modules/projectiles/guns/revolvers.dm (hunk 4 above)
Comment (BraveMole) [reply code/modules/projectiles/guns/revolvers.dm:65]: it's moved to unique action

File: code/modules/projectiles/guns/revolvers.dm [hunk 5]
```diff
@@ -536,3 +534,55 @@
 	burst_delay = 0.1 SECONDS
 	scatter_unwielded = 20
 	damage_mult = 1.05
+
+/obj/item/weapon/gun/revolver/single_action //This town aint big enuf fer the two of us
+	name = "single action revolver"
+	desc = "you should not be seeing this."
+	current_mag = /obj/item/ammo_magazine/internal/revolver/m44
+
+/obj/item/weapon/gun/revolver/single_action/update_icon()
```
Comment (BraveMole) [inline code/modules/projectiles/guns/revolvers.dm:543]: update_icon_state()

File: code/modules/projectiles/gun_system.dm (hunk 3 above)
Comment (BraveMole) [reply code/modules/projectiles/gun_system.dm:367]: standing issue

**Review by BraveMole — CHANGES_REQUESTED** [review]
I kinda fail to see the point of this.

For exemple, switching to hydro in flamers is not cocking? Can you still use keybinds?

rather than moving unique action stuff to cock, why not doing the opposite?

File: code/modules/projectiles/guns/flamer.dm [hunk 6]
```diff
@@ -455,10 +455,7 @@
 	to_chat(user, span_notice("Its hydro cannon contains [reagents.get_reagent_amount(/datum/reagent/water)]/[reagents.maximum_volume] units of water!"))
 
 
-/obj/item/weapon/gun/flamer/big_flamer/marinestandard/unique_action(mob/user)
-	. = ..()
-	if(!.)
-		return
```
Comment (TiviPlus) [inline code/modules/projectiles/guns/flamer.dm:461]: Should be unique action

File: code/modules/projectiles/guns/specialist.dm [hunk 7]
```diff
@@ -101,18 +101,10 @@
...
-		return
+/obj/item/weapon/gun/rifle/sniper/antimaterial/cock(mob/user)
 	if(!targetmarker_primed && !targetmarker_on)
 		return laser_on(user)
-	else
-		return laser_off(user)
-
-/obj/item/weapon/gun/rifle/sniper/antimaterial/cock(mob/user)
-	return
+	return laser_off(user)
```
Comment (TiviPlus) [inline code/modules/projectiles/guns/specialist.dm:107]: ditto

File: code/modules/projectiles/guns/revolvers.dm (hunk 5 above)
Comment (TiviPlus) [reply code/modules/projectiles/guns/revolvers.dm:536]: use update_icon_state here

**Review by TiviPlus — CHANGES_REQUESTED** [review]

**Review by BraveMole — CHANGES_REQUESTED** [review]

File: code/modules/projectiles/guns/revolvers.dm [hunk 8]
```diff
@@ -529,3 +527,55 @@
...
+
+/obj/item/weapon/gun/revolver/single_action/examine(mob/user)
+	. = ..()
+	to_chat(user, "[in_chamber ? "It's primed and ready to fire." : "It is not primed."]")
+
+/obj/item/weapon/gun/revolver/single_action/Fire()
+	. = ..()
+	update_icon()
+
+/obj/item/weapon/gun/revolver/single_action/cock(mob/user)
```
Comment (BraveMole) [inline code/modules/projectiles/guns/revolvers.dm:550]: this one is not returning anything for exemple

File: code/modules/projectiles/guns/specialist.dm [hunk 9]
```diff
@@ -101,18 +101,16 @@
 	remove_overlay(X_LASER_LAYER)
 	return TRUE
 
+/obj/item/weapon/gun/rifle/sniper/antimaterial/cock(mob/user)
+	return TRUE
```
Comment (BraveMole) [inline code/modules/projectiles/guns/specialist.dm:105]: this one is returning true despite doing nothing

**Review by BraveMole — CHANGES_REQUESTED** [review]
Can you explain where is the return for cock (and so for unique action) is used, and why sometimes you return TRUE, sometimes nothing?

**Review by TiviPlus — APPROVED** [review]

**Pull request merged**